### PR TITLE
TC killer

### PR DESCRIPTION
**Pull request by mikusp**
This pull request focuses on providing ability to interrupt running typechecking whenever a new request arrives. Previous approach often resulted in sub-optimal performance and presenting stale results to the user.

On top of that, general robustness of typechecker is improved and there should be fewer situations when observed behaviour is not what was expected.

### CODE REVIEW

File: libs/batch/plugins/luna-empire/src/Empire/Server.hs [hunk 1]
```diff
@@ -115,26 +125,44 @@ prepareStdlib = do
     (cleanup, std) <- Typecheck.createStdlib $ lunaroot <> "/Std/"
     return (std, cleanup)
 
-startTCWorker :: Empire.CommunicationEnv -> Bus ()
-startTCWorker env = liftIO $ do
+killPreviousTC :: Empire.CommunicationEnv -> Maybe (Async Empire.InterpreterEnv) -> IO ()
+killPreviousTC env prevAsync = case prevAsync of
+    Just a -> do
+        res <- Async.poll a
+        case res of
```
Comment (ljkania) [inline libs/batch/plugins/luna-empire/src/Empire/Server.hs]: `Just a -> Async.poll a >>= \case` could save you extra indent and temp variable.

File: libs/batch/plugins/luna-empire/src/Empire/Server/Graph.hs [hunk 2]
```diff
@@ -126,7 +126,7 @@ import qualified ZMQ.Bus.Data.Message                    as Message
 import qualified ZMQ.Bus.EndPoint                        as EP
 import           ZMQ.Bus.Trans                           (BusT (..))
 import qualified ZMQ.Bus.Trans                           as BusT
-
+import qualified System.IO as IO
```
Comment (ljkania) [inline libs/batch/plugins/luna-empire/src/Empire/Server/Graph.hs]: Debug leftover?

File: libs/luna-empire/src/Empire/Commands/Typecheck.hs [hunk 3]
```diff
@@ -44,13 +56,17 @@ import qualified Luna.Project                     as Project
...
+import           Luna.Syntax.Text.Parser.Errors   (Invalids)
+import           OCI.IR.Layout.Typed              (type (>>))
 import           OCI.IR.Name.Qualified            (QualName)
-
-import           System.Directory                     (canonicalizePath)
-import           System.Environment                   (getEnv)
-import qualified Path
+import qualified OCI.Pass.Class                   as Pass (SubPass, eval')
+import qualified OCI.Pass.Manager                 as Pass (PassManager, setAttr, State)
+import           System.Log                                   (DropLogger(..), dropLogs)
```
Comment (ljkania) [inline libs/luna-empire/src/Empire/Commands/Typecheck.hs]: DITTO

File: libs/luna-empire/src/Empire/Commands/Typecheck.hs [hunk 4]
```diff
@@ -1,33 +1,43 @@
...
-import           Control.Monad.State              (execStateT)
+import           Control.Monad.State              (get, execStateT, runStateT)
 import qualified Data.Map                         as Map
 import           Data.Maybe                       (maybeToList)
+import qualified Data.Set                         as Set
 import           Empire.Prelude                   hiding (toList)
 import           Prologue                         (catMaybes, mapping)
 import           System.Directory                 (withCurrentDirectory)
+import           System.Directory                     (canonicalizePath)
+import           System.Environment                   (getEnv)
```
Comment (ljkania) [inline libs/luna-empire/src/Empire/Commands/Typecheck.hs]: Could you fix the indentation?

File: libs/luna-empire/src/Empire/Commands/Typecheck.hs [hunk 5]
```diff
@@ -68,35 +84,37 @@ runInterpreter path imports = runASTOp $ do
...
             res     <- Interpreter.interpret' imports . IR.unsafeGeneralize $ bodyRef
-            result  <- liftIO $ withCurrentDirectory (maybe (takeDirectory path) Path.toFilePath rootPath) $ runIO $ runError $ execStateT res def
-            case result of
-                Left e  -> return Nothing
-                Right r -> return $ Just r
+            mask_ $ do
+                result  <- liftIO $ withCurrentDirectory (maybe (takeDirectory path) Path.toFilePath rootPath) $ runIO $ runError $ execStateT res def
+                case result of
+                    Left e  -> return Nothing
+                    Right r -> return $ Just r
```
Comment (ljkania) [inline libs/luna-empire/src/Empire/Commands/Typecheck.hs]: `return $ either Nothing (Just r) result` ?

File: libs/luna-empire/src/Empire/Commands/Typecheck.hs [hunk 6]
```diff
@@ -168,26 +187,31 @@ getCurrentScope imports file = do
...
-            scope      <- (if recompute then recomputeCurrentScope else getCurrentScope) imports file
-            let imps = unionImports (flattenScope $ Scope std) scope
-            zoom graph $ flip (zoomBreadcrumb' br) (return ()) $ do
-                runTC imps
-                updateNodes  loc
-                {-updateMonads loc-}
-                when interpret $ do
-                    scope <- runInterpreter file imps
-                    traverse_ (updateValues loc) scope
+            void $ mask_ $ recomputeCurrentScope imports file
```
Comment (ljkania) [inline libs/luna-empire/src/Empire/Commands/Typecheck.hs]: It does not need `do`, does it?

File: libs/luna-empire/src/Empire/Commands/Typecheck.hs (hunk 5 above)
Comment (mikusp) [reply libs/luna-empire/src/Empire/Commands/Typecheck.hs]: @kustosz I see we're ignoring the exception string here, was it done on purpose?

File: libs/luna-empire/src/Empire/Commands/Typecheck.hs (hunk 6 above)
Comment (mikusp) [reply libs/luna-empire/src/Empire/Commands/Typecheck.hs]: While technically it doesn't, I think it looks ok here to have consistency with next clause

File: libs/batch/plugins/luna-empire/src/Empire/Server.hs (hunk 1 above)
Comment (mikusp) [reply libs/batch/plugins/luna-empire/src/Empire/Server.hs]: done

File: libs/batch/plugins/luna-empire/src/Empire/Server/Graph.hs (hunk 2 above)
Comment (mikusp) [reply libs/batch/plugins/luna-empire/src/Empire/Server/Graph.hs]: yup

File: libs/batch/plugins/luna-empire/src/Empire/Server/Graph.hs [hunk 7]
```diff
@@ -487,9 +486,15 @@ handleSubstitute = modifyGraph defInverse action replyResult where
     action req@(Substitute.Request location diffs) = do
         let file = location ^. GraphLocation.filePath
         prevImports <- Graph.getAvailableImports location
-        res         <- withDefaultResultTC location $ Graph.substituteCodeFromPoints file diffs
+        res         <- withDefaultResult location $ Graph.substituteCodeFromPoints file diffs
         newImports  <- Graph.getAvailableImports location
-        return $ Substitute.Result res $ if Set.fromList prevImports == Set.fromList newImports then def else return newImports
+        let importChange = if Set.fromList prevImports == Set.fromList newImports then Nothing else Just newImports
+        if isJust importChange then do
+            Graph.typecheckWithRecompute (GraphLocation file def)
```
Comment (kustosz) [inline libs/batch/plugins/luna-empire/src/Empire/Server/Graph.hs]: Do we really need both? The whole point of `typecheckWithRecompute` was to run exactly the operation of running toplevel TC, so why is this needed?

File: libs/luna-empire/src/Empire/ASTOp.hs [hunk 8]
```diff
@@ -245,36 +252,51 @@ runTypecheck imports = do
...
+
+tcInit = do
+    Pass.setAttr (getTypeDesc @WorldExpr)                 $ error "Data not provided: WorldExpr"
+    Pass.setAttr (getTypeDesc @UnitLoader.UnitsToLoad)    $ error "Data not provided: UnitsToLoad"
+    Pass.setAttr (getTypeDesc @UnitLoader.SourcesManager) $ error "Data not provided: SourcesManager"
+    Pass.setAttr (getTypeDesc @UnitSet)                   $ error "Data not provided: UnitSet"
+    Pass.setAttr (getTypeDesc @Invalids)                  $ (mempty :: Invalids)
+    initNameGen
+
+extractImportedModules unit = do
```
Comment (kustosz) [inline libs/luna-empire/src/Empire/ASTOp.hs]: Type it, no cheating!

File: libs/luna-empire/src/Empire/ASTOp.hs [hunk 9]
```diff
@@ -245,36 +252,51 @@ runTypecheck imports = do
...
     put $ newG & Graph.ast .~ AST st passSt
 
+evalTC g ir pmState = flip runStateT g
+                    . withVis
+                    . dropLogs
+                    . DepState.evalDefStateT @Cache
+                    . flip evalIRBuilder ir
+                    . flip evalPassManager pmState
+
+tcInit = do
```
Comment (kustosz) [inline libs/luna-empire/src/Empire/ASTOp.hs]: Type it, no cheating!

File: libs/luna-empire/src/Empire/Commands/Graph.hs [hunk 10]
```diff
@@ -205,6 +206,8 @@ addImports loc@(GraphLocation file _) modulesToImport = do
         let newImports = map (\i -> Text.concat ["import ", i, "\n"]) neededImports
         return $ Text.concat $ newImports ++ [code]
     reloadCode loc newCode
+    typecheckWithRecompute (GraphLocation file def)
```
Comment (kustosz) [inline libs/luna-empire/src/Empire/Commands/Graph.hs]: again, why recompute after recompute?

File: libs/luna-empire/src/Empire/Commands/Typecheck.hs (hunk 5 above)
Comment (kustosz) [reply libs/luna-empire/src/Empire/Commands/Typecheck.hs]: It was done because I had no better idea what to do with this error. This says "the whole computation ended with an exception", without even a hint as to what node caused it. At some point it will be probably displayed somewhere, but is ignored for a long time now.

File: libs/batch/plugins/luna-empire/src/Empire/Server.hs [hunk 11]
```diff
@@ -115,26 +126,42 @@ prepareStdlib = do
     (cleanup, std) <- Typecheck.createStdlib $ lunaroot <> "/Std/"
     return (std, cleanup)
 
-startTCWorker :: Empire.CommunicationEnv -> Bus ()
-startTCWorker env = liftIO $ do
+killPreviousTC :: Empire.CommunicationEnv -> Maybe (Async Empire.InterpreterEnv) -> IO ()
+killPreviousTC env prevAsync = case prevAsync of
+    Just a -> Async.poll a >>= \case
```
Comment (piotrMocz) [inline libs/batch/plugins/luna-empire/src/Empire/Server.hs]: Actually, you could simplify it even further by writing:
```
killPreviousTC env = \case
    Nothing -> return ()
    Just prevAsync -> Async.poll prevAsync >>= \case [...]
```
Oh and I like to put the significantly-shorter branch first, so that you don't get a lone `_ -> return ()` at the end of the function. Just a matter of preference, though.

File: libs/luna-empire/src/Empire/ASTOp.hs [hunk 12]
```diff
@@ -245,36 +253,58 @@ runTypecheck imports = do
         return (st, passSt)
     put $ newG & Graph.ast .~ AST st passSt
 
+evalTC :: ClsGraph
+       -> IR
+       -> State (Pass.PassManager (IRBuilder (DepState.StateT Cache (Logger DropLogger (Vis.VisStateT (StateT ClsGraph IO))))))
+       -> Pass.PassManager (IRBuilder (DepState.StateT Cache (Logger DropLogger (Vis.VisStateT (StateT ClsGraph IO))))) a
```
Comment (piotrMocz) [inline libs/luna-empire/src/Empire/ASTOp.hs]: o.0
